include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -1073,11 +1086,11 @@ def UnionCreateOp : MooreOp<"union_create"> {
     ```
     See IEEE 1800-2017 § 7.3 "Unions"
   }];
-  let arguments = (ins StrAttr:$unionName, UnpackedType:$input);
-  let results = (outs UnpackedType:$result);
+  let arguments = (ins StrAttr:$name, UnpackedType:$input);
+  let results = (outs AnyUnionType:$result);
   let assemblyFormat = [{
-  $unionName `,` $input attr-dict `:`
-  type($input) `->` type($result)
+    `` custom<ImplicitSSAName>($name) $input attr-dict
+    `:` type($input) `->` type($result)
```

Comment:
Same as with `struct_create`: the op only creates an SSA union value in the IR, but that doesn't have a name itself. Names would be provided by a `moore.variable` op that uses that SSA union value as an initial value, or an assignment that stores this SSA value into a variable.

However, I agree that you _do_ need a name here: the name of the union variant that should be used to create the union value. For example, a union `union<a: i42, b: i9001>` has two variants, "a" and "b". When you construct such a union, you have to pick which one of the variants you are choosing. The options would be:
```mlir
moore.union_create %0, "a" : i42 -> union<a: i42, b: i9001>
moore.union_create %1, "b" : i9001 -> union<a: i42, b: i9001>
```
So depending on which `variantName` the operation picks, the operand has to have a type corresponding to that variant in the union type. ("a" needs i42, "b" needs i9001).

You can use `TypesMatchWith<...>` constraints again to ensure that the type of the operand `%0` or `%1` matches the type of the field `$variantName` in `$result`.

The `OpAsmOpInterface` is then no longer needed.